client/packages/system/src/Stock/Components/StockLineEditModal.tsx

```diff
@@ -21,25 +21,31 @@ interface StockLineEditModalProps {
 }
 
 interface UseDraftStockLineControl {
-  draft: StockLineRowFragment;
+  draft: StockLineRowFragment | null;
   onUpdate: (patch: Partial<StockLineRowFragment>) => void;
   onSave: () => Promise<void>;
   isLoading: boolean;
 }
 
 const useDraftStockLine = (
-  seed: StockLineRowFragment
+  seed: StockLineRowFragment | null
 ): UseDraftStockLineControl => {
-  const [stockLine, setStockLine] = useState<StockLineRowFragment>(() => ({
-    ...seed,
-  }));
+  const [stockLine, setStockLine] = useState<StockLineRowFragment | null>(() =>
+    seed
+      ? {
+          ...seed,
+        }
+      : null
+  );
   const { mutate, isLoading } = useStock.line.update();
 
   const onUpdate = (patch: Partial<StockLineRowFragment>) => {
-    setStockLine({ ...stockLine, ...patch });
+    if (stockLine) setStockLine({ ...stockLine, ...patch });
   };
 
-  const onSave = async () => mutate(stockLine);
+  const onSave = async () => {
+    if (stockLine) mutate(stockLine);
+  };
```

Comment:
with
```
interface UseDraftStockLineControl {
  draft: StockLineRowFragment;
  onUpdate: (patch: Partial<StockLineRowFragment>) => void;
  onSave: () => Promise<void>;
  isLoading: boolean;
}
```

the hook can be a little easier to use like this

```
const useDraftStockLine = (
  seed: StockLineRowFragment | null
): UseDraftStockLineControl => {
  const initialState = seed ? { ...seed } : ({} as StockLineNode);
  const [draft, setDraft] = useState(initialState);
  const { mutate, isLoading } = useStock.line.update();

  const onUpdate = (patch: Partial<StockLineRowFragment>) => {
    setDraft({ ...draft, ...patch });
  };

  const onSave = async () => {
    mutate(draft);
  };

  return {
    draft,
    onUpdate,
    onSave,
    isLoading,
  };
};
```

and below becomes:

```
  if (!stockLine) return null;
```

this is more functionally similar to the original - could also have the `draft` as `StockLineNode | null` if you prefer not to have the empty object cast as a type.. that's a little off I know but it does allow the patching of the draft, rather than preventing any updates as the null version does.

Reply:
Previously, there was this:
```ts
  if (!stockLine) return null;
  const { draft, onUpdate, onSave } = useDraftStockLine(stockLine);
```

Isn't it the same? i.e. previously, when there was no `stockLine` nothing gets rendered. And with my change, if there isn't a stockLine there is no draft and nothing gets rendered...

Reply:
I thought it might be easier with: https://github.com/openmsupply/open-msupply/compare/react-hook-linting...react-hook-linting(Not-null-stockline).

I am a big fan of strict type safety, maybe using `useEditModal` with optional entity was the original issue ?

Reply:
Excellent! merged this in. Thanks

Reply:
> Isn't it the same?

It's the changes to not update or save if the draft is null that is the functional change

```
if (stockLine) setStockLine({ ...stockLine, ...patch });
```

which, in reality, will behave the same because the component isn't rendered if the draft is null, so the updates aren't called anyway. As well as that, the stockline won't save if the draft isn't populated, as the ID will be empty. So yes, what I'm asking for won't have any impact.. but it is neater to remove the null and have stricter typing..

And with your change @andreievg - what the draft object pattern is doing everywhere is creating a new object to mutate and update rather than the directly mutating, which is what the `useState<StockLineRowFragment>(seed);` will do.

Once again, realistically this will have little impact - since the `seed` object is the data returned by the API call, and react query won't be caching based on the object itself.. It still feels nicer to keep to the immutable pattern and create a new object to mutate.

Reply:
thanks, I changed it back what it was previously doing: `useState<StockLineRowFragment>({ ...seed })`

Reply:
@mark-prins 
> And with your change @andreievg - what the draft object pattern is doing everywhere is creating a new object to mutate and update rather than the directly mutating, which is what the useState<StockLineRowFragment>(seed); will do

Sorry I can't quite figure out why it's not safe, it's initial (starting) value, and we would always shallow copy when setting it again, so initial value is never edited again ¯\_(ツ)_/¯

Reply:
Think, if you just use `useState(seed)`  useState just uses the seed object as it it, and if seed gets mutated somewhere else the state doesn't get updated. However, even if you do a shallow copy like `{...seed}` this could still happen for child objects of seed since they still point to the original object... Probably safe here though but good practice to do it anyway?